crates/tako/src/internal/datasrv/local_client.rs

```diff
@@ -48,32 +58,92 @@ impl LocalDataClient {
         })
     }
 
-    pub async fn put_data_object(
+    pub async fn put_data_object_from_file(
         &mut self,
         data_id: OutputId,
         mime_type: String,
-        data: Vec<u8>,
+        path: &Path,
     ) -> crate::Result<()> {
-        let message = FromLocalDataClientMessage::PutDataObject {
+        log::debug!(
+            "Uploading file {} as data_obj={} (mime_type={})",
+            path.display(),
             data_id,
-            data_object: DataObjectRef::new(DataObject::new(mime_type, data)),
-        };
-        self.send_message(message).await?;
+            &mime_type
+        );
+        let mut file = File::open(path)?;
+        let size = file.metadata()?.len();
+        let mut buffer = Vec::new();
+        buffer.resize(UPLOAD_CHUNK_SIZE, 0);
```

Comment:
I think that 32 MiB is too large of a buffer. I would use something smaller, e.g. 64/128/256 KiB. You want the buffer to be small enough to stay in cache, so that the following I/O has faster memory throughput. That beats the larger number of syscalls made, the data is probably buffered in the OS pages anyway (https://www.evanjones.ca/read-write-buffer-size.html).